File: worker/src/paid_egress.py
from __future__ import annotations

import json
import os
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

import requests

from src.timezone import BUSINESS_TIMEZONE


class PaidEgressBlocked(RuntimeError):
    pass


Provider = str
Context = str

WORKER_ROOT = Path(__file__).resolve().parent.parent
USAGE_LOG = WORKER_ROOT / "logs" / "provider_usage_events.jsonl"


def default_context() -> Context:
    return os.environ.get("PAID_EGRESS_CONTEXT", "scheduled_pipeline")


def manual_enrich_context(company_id: str) -> Context:
    return f"manual_enrich:{company_id}"


def _env_flag(name: str, default: bool) -> bool:
    raw = os.environ.get(name)
    if raw is None or raw.strip() == "":
        return default
    return raw.strip().lower() in {"1", "true", "yes", "on"}


def _provider_prefix(provider: Provider) -> str:
    return "APOLLO" if provider == "apollo" else "CONTACTOUT"


def _provider_enabled(provider: Provider) -> bool:
    if not _env_flag("PAID_EGRESS_ENABLED", True):
        return False
    if not _env_flag(f"{_provider_prefix(provider)}_EGRESS_ENABLED", True):
        return False
    return _env_flag(f"{_provider_prefix(provider)}_PAID_EGRESS_ENABLED", True)


# Internal safety cap on estimated credits per business day — a guardrail so a
# bug can't drain the provider balance, NOT the provider's real balance.
# ContactOut default = daily enrich quota (25) x contacts (3) x credits (2).
def _daily_cap(provider: Provider) -> int:
    raw = os.environ.get(f"{_provider_prefix(provider)}_DAILY_CREDIT_CAP")
    if raw and raw.strip().isdigit():
        return int(raw)
    return 200 if provider == "apollo" else 150


def _today() -> str:
    # Business-day window (midnight ET) — matches the CRM's cap window; a UTC
    # day would roll at 8 PM ET and charge evening usage to the next day.
    return datetime.now(BUSINESS_TIMEZONE).date().isoformat()


def record_provider_usage(
    provider: Provider,
    endpoint: str,
    context: Context,
    *,
    records_returned: int = 0,
    estimated_cost: int = 0,
    blocked: bool = False,
    metadata: dict[str, Any] | None = None,
) -> None:
    USAGE_LOG.parent.mkdir(parents=True, exist_ok=True)
    event = {
        "created_at": datetime.now(timezone.utc).isoformat(),
        "date": _today(),
        "provider": provider,
        "endpoint": endpoint,
        "egress_context": context,
        "trigger_source": context.split(":", 1)[0],
        "records_returned": records_returned,
        "estimated_cost": estimated_cost,
        "blocked": blocked,
        "metadata": metadata or {},
    }
    with USAGE_LOG.open("a", encoding="utf-8") as handle:
        handle.write(json.dumps(event, sort_keys=True) + "\n")
    _post_provider_usage(event)


def _post_provider_usage(event: dict[str, Any]) -> None:
    base = (os.environ.get("CRM_API_URL") or "").rstrip("/")
    key = os.environ.get("CRM_API_KEY", "")
    if not base or not key:
        return
    try:
        requests.post(
            f"{base}/api/pipeline/usage",
            headers={
                "Content-Type": "application/json",
                "Authorization": f"Bearer {key}",
            },
            json=event,
            timeout=5,
        )
    except requests.RequestException:
        return


def _daily_usage(provider: Provider) -> int:
    if not USAGE_LOG.exists():
        return 0
    total = 0
    today = _today()
    try:
        with USAGE_LOG.open("r", encoding="utf-8") as handle:
            for line in handle:
                try:
                    event = json.loads(line)
                except json.JSONDecodeError:
                    continue
                if (
                    event.get("date") == today
                    and event.get("provider") == provider
                    and not event.get("blocked")
                ):
                    total += int(event.get("estimated_cost") or 0)
    except OSError:
        return 0
    return total


def assert_paid_egress_allowed(
    provider: Provider,
    endpoint: str,
    *,
    context: Context | None = None,
    estimated_cost: int = 1,
    metadata: dict[str, Any] | None = None,
) -> Context:
    ctx = context or default_context()
    if not _provider_enabled(provider):
        record_provider_usage(
            provider,
            endpoint,
            ctx,
            blocked=True,
            metadata={**(metadata or {}), "reason": "provider_disabled"},
        )
        raise PaidEgressBlocked(f"{provider} paid egress disabled for {endpoint}")

    if not ctx.startswith("manual_enrich:"):
        record_provider_usage(
            provider,
            endpoint,
            ctx,
            blocked=True,
            metadata={**(metadata or {}), "reason": "non_manual_context"},
        )
        raise PaidEgressBlocked(f"{provider} paid egress blocked for {ctx} ({endpoint})")

    cap = _daily_cap(provider)
    used = _daily_usage(provider)
    if used + estimated_cost > cap:
        record_provider_usage(
            provider,
            endpoint,
            ctx,
            blocked=True,
            metadata={
                **(metadata or {}),
                "reason": "daily_cap_reached",
                "cap": cap,
                "used_today": used,
            },
        )
        raise PaidEgressBlocked(
            f"{provider} daily safety cap reached — {used}/{cap} estimated credits "
            f"used since midnight ET. This is the worker's own guardrail, not the "
            f"{provider} balance; set {_provider_prefix(provider)}_DAILY_CREDIT_CAP "
            f"to raise it. Resets at midnight ET."
        )

    return ctx

File: src/lib/paid-egress.ts
import { and, eq, gte, notInArray, sql } from "drizzle-orm";
import { db } from "@/lib/db";
import { providerUsageEvents } from "@/lib/db/schema";
import { businessDayStartUtc } from "@/lib/timezone";

/**
 * `serpapi` bills per search rather than per revealed record, and the Mac
 * worker's Google Jobs scrape already posts `provider = 'serpapi'` usage
 * events here. Bringing it into this union means CRM-side SerpApi callers get
 * the same daily hard stop and the same audit trail as Apollo and ContactOut,
 * and the daily total covers every SerpApi consumer the app can see rather
 * than one of them.
 */
export type PaidProvider = "apollo" | "contactout" | "serpapi";

export type PaidEgressContext =
  | `manual_enrich:${string}`
  | "health_check"
  | "manual_script"
  | "automated_scrape"
  | "scheduled_pipeline";

export type PaidEgressMetadata = {
  companyId?: string | null;
  contactId?: string | null;
  recordsReturned?: number;
  estimatedCost?: number;
  metadata?: Record<string, unknown>;
};

export class PaidEgressBlockedError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "PaidEgressBlockedError";
  }
}

export function manualEnrichContext(companyId: string): PaidEgressContext {
  return `manual_enrich:${companyId}`;
}

export function isManualEnrichContext(
  context: PaidEgressContext | null | undefined,
): context is `manual_enrich:${string}` {
  return Boolean(context?.startsWith("manual_enrich:"));
}

function triggerSource(context: PaidEgressContext): string {
  return context.split(":")[0] || context;
}

const ENV_PREFIX: Record<PaidProvider, string> = {
  apollo: "APOLLO",
  contactout: "CONTACTOUT",
  serpapi: "SERPAPI",
};

function providerEnvPrefix(provider: PaidProvider): string {
  return ENV_PREFIX[provider];
}

function envFlag(value: string | undefined, defaultValue: boolean): boolean {
  if (value == null || value.trim() === "") return defaultValue;
  return ["1", "true", "yes", "on"].includes(value.trim().toLowerCase());
}

export function paidEgressEnabled(provider: PaidProvider): boolean {
  const prefix = providerEnvPrefix(provider);
  if (!envFlag(process.env.PAID_EGRESS_ENABLED, true)) return false;
  if (!envFlag(process.env[`${prefix}_EGRESS_ENABLED`], true)) return false;
  return envFlag(process.env[`${prefix}_PAID_EGRESS_ENABLED`], true);
}

/**
 * Internal safety cap on estimated credits per business day — a guardrail so
 * a bug can't drain the provider balance. NOT the provider's real balance.
 * ContactOut default follows the credit-governance formula in the playbook:
 * daily enrich quota (25) × contacts per company (3) × credits per contact (2).
 *
 * SerpApi's unit is a search, not a credit, and the plan is monthly: the
 * default is the guarded monthly budget spread evenly over a month
 * (SERPAPI_MONTHLY_PLAN 15000 × SERPAPI_BUDGET_PCT 0.8 ÷ 30 ≈ 400). This cap
 * is shared with the worker's Google Jobs scrape, which posts its searches to
 * the same table — deliberately, so one number bounds all SerpApi spend the
 * app can see.
 */
const DEFAULT_DAILY_CAP: Record<PaidProvider, number> = {
  apollo: 200,
  contactout: 150,
  serpapi: 400,
};

export function providerDailyCap(provider: PaidProvider): number {
  const prefix = providerEnvPrefix(provider);
  const raw = process.env[`${prefix}_DAILY_CREDIT_CAP`];
  const parsed = raw ? Number.parseInt(raw, 10) : NaN;
  if (Number.isFinite(parsed) && parsed >= 0) return parsed;
  return DEFAULT_DAILY_CAP[provider];
}

/**
 * Endpoints the provider documents as costing nothing, so they can never
 * consume a credit cap no matter what an older row happened to record.
 *
 * Apollo's People API Search is explicitly "Credit usage: 0 credits"
 * (https://docs.apollo.io/docs/api-pricing). It used to be logged at 1 credit
 * a call, and discovery makes up to four per company, so a normal batch could
 * exhaust a day's budget on calls that were never billed. Excluding the
 * endpoint here — rather than only fixing what new rows record — also
 * discounts the rows already written under the old accounting.
 *
 * ContactOut's people/search is NOT free: it is sent with `reveal_info: true`
 * and bills per revealed profile, so it stays counted.
 *
 * SerpApi has none: it bills every successful search regardless of how many
 * results come back, so no endpoint of theirs is free.
 */
const ZERO_COST_ENDPOINTS: Record<PaidProvider, readonly string[]> = {
  apollo: ["mixed_people/api_search"],
  contactout: [],
  serpapi: [],
};

export function endpointConsumesCredits(
  provider: PaidProvider,
  endpoint: string,
): boolean {
  return !ZERO_COST_ENDPOINTS[provider].includes(endpoint);
}

export async function dailyUsage(provider: PaidProvider): Promise<number> {
  const free = ZERO_COST_ENDPOINTS[provider];
  const [row] = await db
    .select({
      total: sql<number>`coalesce(sum(${providerUsageEvents.estimatedCost}), 0)`,
    })
    .from(providerUsageEvents)
    .where(
      and(
        eq(providerUsageEvents.provider, provider),
        eq(providerUsageEvents.blocked, false),
        // Business-day window (midnight ET) — a UTC window rolls at 8 PM ET
        // and charges evening usage against the next day's budget.
        gte(providerUsageEvents.createdAt, businessDayStartUtc()),
        ...(free.length
          ? [notInArray(providerUsageEvents.endpoint, [...free])]
          : []),
      ),
    );
  return Number(row?.total ?? 0);
}

export async function recordProviderUsageEvent(
  provider: PaidProvider,
  endpoint: string,
  context: PaidEgressContext,
  details: PaidEgressMetadata & { blocked?: boolean } = {},
): Promise<void> {
  try {
    await db.insert(providerUsageEvents).values({
      provider,
      endpoint,
      egressContext: context,
      triggerSource: triggerSource(context),
      companyId: details.companyId ?? null,
      contactId: details.contactId ?? null,
      recordsReturned: details.recordsReturned ?? 0,
      estimatedCost: details.estimatedCost ?? 0,
      blocked: details.blocked ?? false,
      metadata: details.metadata,
    });
  } catch (err) {
    console.error("Failed to log provider usage event", err);
  }
}

export async function assertPaidEgressAllowed(
  provider: PaidProvider,
  endpoint: string,
  context: PaidEgressContext | null | undefined,
  details: PaidEgressMetadata = {},
): Promise<void> {
  const normalizedContext = context ?? "automated_scrape";
  const estimatedCost = details.estimatedCost ?? 1;

  if (!paidEgressEnabled(provider)) {
    await recordProviderUsageEvent(provider, endpoint, normalizedContext, {
      ...details,
      estimatedCost: 0,
      blocked: true,
      metadata: {
        ...(details.metadata ?? {}),
        reason: "provider_disabled",
      },
    });
    throw new PaidEgressBlockedError(
      `${provider} paid egress is disabled for ${endpoint}`,
    );
  }

  if (!isManualEnrichContext(normalizedContext)) {
    await recordProviderUsageEvent(provider, endpoint, normalizedContext, {
      ...details,
      estimatedCost: 0,
      blocked: true,
      metadata: {
        ...(details.metadata ?? {}),
        reason: "non_manual_context",
      },
    });
    throw new PaidEgressBlockedError(
      `${provider} paid egress blocked for ${normalizedContext} (${endpoint})`,
    );
  }

  const cap = providerDailyCap(provider);
  if (cap >= 0) {
    const used = await dailyUsage(provider);
    if (used + estimatedCost > cap) {
      await recordProviderUsageEvent(provider, endpoint, normalizedContext, {
        ...details,
        estimatedCost: 0,
        blocked: true,
        metadata: {
          ...(details.metadata ?? {}),
          reason: "daily_cap_reached",
          cap,
          usedToday: used,
        },
      });
      throw new PaidEgressBlockedError(
        `${provider} daily safety cap reached — ${used}/${cap} estimated credits used since midnight ET. ` +
          `This is the app's own guardrail, not your ${provider} balance; ` +
          `set ${providerEnvPrefix(provider)}_DAILY_CREDIT_CAP on Vercel to raise it. Resets at midnight ET.`,
      );
    }
  }
}
